golang/src/stream_manager.rs

```diff
@@ -0,0 +1,354 @@
+use libc::c_char;
+use pravega_client::client_factory::ClientFactory;
+use pravega_client::event::reader_group::ReaderGroupConfigBuilder;
+use pravega_client_config::ClientConfigBuilder;
+use pravega_client_shared::*;
+use std::ffi::CStr;
+use std::panic::{catch_unwind, AssertUnwindSafe};
+use std::ptr;
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use crate::error::{clear_error, set_error};
+use crate::memory::Buffer;
+use crate::reactor::Incoming;
+use crate::stream_reader_group::StreamReaderGroup;
+use crate::stream_writer::StreamWriter;
+use crate::reactor::*;
+pub struct StreamManager {
+    cf: ClientFactory,
+    sender: UnboundedSender<Incoming>
+}
+
+impl StreamManager {
+    fn new(
+        controller_uri: &str,
+    ) -> Self {
+        // make sure the error number is 0 during startup
+        clear_error();
+
+        let mut builder = ClientConfigBuilder::default();
+        builder
+            .controller_uri(controller_uri)
+            .is_auth_enabled(false)
+            .is_tls_enabled(false);
+        let config = builder.build().expect("creating config");
+        let client_factory = ClientFactory::new(config.clone());
+
+        // start reactor
+        let handle = client_factory.runtime_handle();
+        let (tx, rx) = unbounded_channel();
+        handle.spawn(Reactor::run(rx));
```

Comment:
Humm.... I don't think this is what we want. 
Part of the idea was that if we have a go thread we can use it to call back into go for free. Here it is spawning a task on the generic background threadpool so it won't use the go thread. It would be better if we had a `run` method or something that go would call from it's wrapper and then we could capture the thread.

Reply:
@tkaitchuck, I removed the reactor because the operations shouldn't be handled in one task which will serialize the operations. Also it won't need to retrieve identifiers from a Rust channel at Golang side as `ackOperationDone` would notify the Golang channel when the resource is ready at the Rust side.  I reimplement get_segment_slice as https://github.com/vangork/pravega-client-rust/blob/c97f91e077639bf56f8969cc572596155d721c47/golang/src/stream_reader.rs#L39-L52. Would like to have your comment again.